ns_pipelines/pipeline.py

```diff
@@ -0,0 +1,238 @@
+from datetime import datetime
+import json
+import hashlib
+from abc import ABC, abstractmethod
+from functools import reduce
+from pathlib import Path
+from typing import Dict, Any, List, Union
+
+
+INPUTS = [
+    "text",
+    "coordinates",
+    "metadata",
+]
+
+RAW_INPUTS = [
+    "raw.html",
+    "raw.xml",
+    "raw.tables",
+    "raw.tables_xml",
+]
+
+
+def deep_getattr(obj: Any, attr_path: str, default: Any = None) -> Any:
+    try:
+        return reduce(getattr, attr_path.split('.'), obj)
+    except AttributeError:
+        return default
+
+
+class FileManager:
+    """Utility class for file handling operations."""
+
+    @staticmethod
+    def calculate_md5(file_path: Path) -> str:
+        """Calculate MD5 hash of a file."""
+        with file_path.open('r') as f:
+            file_contents = f.read()
+        return hashlib.md5(file_contents.encode()).hexdigest()
+
+    @staticmethod
+    def load_json(file_path: Path) -> Dict:
+        """Load JSON from a file."""
+        with file_path.open('r') as f:
+            return json.load(f)
+
+    @staticmethod
+    def write_json(file_path: Path, data: Dict):
+        """Write JSON to a file."""
+        with file_path.open('w') as f:
+            json.dump(data, f)
+
+    @staticmethod
+    def get_next_available_dir(base_path: Path) -> Path:
+        """Find the next available directory by appending numbers (-1, -2, etc.) if necessary."""
+        counter = 1
+        new_path = base_path
+        while new_path.exists():
+            new_path = base_path.with_name(f"{base_path.name}-{counter}")
+            counter += 1
+        return new_path
+
+
+class Pipeline(ABC):
+    """Abstract pipeline class for processing data."""
+
+    _hash_args: List[str] = []
```

Comment:
Another idea here is that instead of defining which arguments are logged on not logged, we differentiate between the two types of arguments and those are set as class attributes.


i.e. 
```
__arguments = [arg_will_be_logged, arg 2],
__runtime_args = [argumentnot_logged]
```

I think we want to differentiate between a few things: How to specific which arguments are allowable/expected by a pipeline, which are required by a pipeline (and perhaps have no default value so they must be specific), and which are to be logged. Perhaps, there can be a default value for which are to be logged as well. 

Relatedly: for setting Pipeline arguments, do we want to just procesed them in the baseclass `Pipeline.__init__`, or do we want to have pipeline subclass the `def __init__` method? and class __super__?

Finally: is there any scenario we would not want to hash `["_inputs", "_input_sources"]`? So perhaps that should always be part of the hash, and cannot be disabled, and the PIpeline specific arguments are added to that hash.

Reply:
we are differentiating runtime_args to be passed/defined in the run function, and the hash arguments are automatically extracted from the __init__ function of the pipeline. Writers of pipeline will need to be cognizant of only including pipeline configuration changing arguments in the __init__ and using runtime arguments (n_cpus, n_threads) in the run function.